#' Raw Per-Protein Embeddings For \emph{Escherichia coli}
#'
#' Raw per-protein embeddings generated by UniProt using the ProtT5 model
#' (prottrans_t5_xl_u50) for the reference proteome of \emph{Escherichia coli}.
#'
#' The embedding matrix for several species can be downloaded
#' \href{https://www.uniprot.org/help/downloads#embeddings}{here}. Protein
#' sequences longer than 12k residues were excluded due to computational
#' limitations.
#'
#' @format A data frame with 4,401 rows and 1024 total columns. In the index are
#'     UniProt identifiers for each protein, and columns are labelled V1, V2,
#'     ..., V1024 to represent each dimension of the embedding.
#'
#' @source UniProt Consortium.

#' @references
#' Bateman, A., Martin, M.-J., Orchard, S., Magrane, M., Ahmad, S., Alpi, E.,
#' Bowler-Barnett, E. H., Britto, R., Bye-A-Jee, H., Cukura, A., Denny, P.,
#' Dogan, T., Ebenezer, T., Fan, J., Garmiri, P., da Costa Gonzales, L. J.,
#' Hatton-Ellis, E., Hussein, A., Ignatchenko, A., & Insana, G. (2022).
#' UniProt: the Universal Protein Knowledgebase in 2023. \emph{Nucleic Acids
#' Research}, 51(D1). \href{https://doi.org/10.1093/nar/gkac1052}{Link}.
#'
#' @examples
#' \dontrun{
#'  eColiEmbeddingMatrix
#' }
"eColiEmbeddingMatrix"

#' Paralog Mapping for \emph{Escherichia coli}
#'
#' A one-to-many paralog mapping for \emph{Escherichia coli} generated from the
#' Orthologous Matrix (OMA) Browser.
#'
#' Paralog mapping was retrieved using the
#' \href{https://omabrowser.org/oma/genomePW/}{Genome Pair View} tool with both
#' species as \strong{Escherichia coli str. K-12 substr. MG1655 (ECOLI)}.
#' Identifiers were converted to UniProt using the Identifier Mapping file for
#' download on the OMA Browser. Processing was done to remove rows where either
#' protein was not present in the eColiEmbeddingMatrix to reduce the size of the
#' file.
#'
#' @format A data frame with 7,550 rows and 2 total columns. Columns Protein 1
#'     and Protein 2 are UniProt identifiers, and each row represents one
#'     paralogous pair.
#'
#' @source Orthologous Matrix (OMA) Browser
#'
#' @references
#' Altenhoff, A. M., Glover, N. M., Train, C. M., Kaleb, K., Warwick Vesztrocy,
#' A., Dylus, D., de Farias, T. M., Zile, K., Stevenson, C., Long, J., Redestig,
#' H., Gonnet, G. H., & Dessimoz, C. (2018). The OMA orthology database in 2018:
#' retrieving evolutionary relationships among all domains of life through
#' richer web and programmatic interfaces. Nucleic Acids Research, 46(D1),
#' D477-D485. \href{https://pubmed.ncbi.nlm.nih.gov/29106550/}{Link}.
#'
#' @examples
#' \dontrun{
#'  eColiParalogMapping
#' }
"eColiParalogMapping"



#' Raw Per-Protein Embeddings For \emph{Severe Acute Respiratory Syndrome
#' Coronavirus 2 (SARS CoV-2)}
#'
#' Raw per-protein embeddings generated by UniProt using the ProtT5 model
#' (prottrans_t5_xl_u50) for the reference proteome of \emph{SARS CoV-2}.
#'
#' The embedding matrix for several species can be downloaded
#' \href{https://www.uniprot.org/help/downloads#embeddings}{here}. Protein
#' sequences longer than 12k residues were excluded due to computational
#' limitations. Only embeddings for non-structural proteins were generated.
#'
#' @format A data frame with 17 rows and 1024 total columns. In the index are
#'     UniProt identifiers for each protein, and columns are labelled V1, V2,
#'     ..., V1024 to represent each dimension of the embedding.
#'
#' @source UniProt Consortium.

#' @references
#' Bateman, A., Martin, M.-J., Orchard, S., Magrane, M., Ahmad, S., Alpi, E.,
#' Bowler-Barnett, E. H., Britto, R., Bye-A-Jee, H., Cukura, A., Denny, P.,
#' Dogan, T., Ebenezer, T., Fan, J., Garmiri, P., da Costa Gonzales, L. J.,
#' Hatton-Ellis, E., Hussein, A., Ignatchenko, A., & Insana, G. (2022).
#' UniProt: the Universal Protein Knowledgebase in 2023. \emph{Nucleic Acids
#' Research}, 51(D1). \href{https://doi.org/10.1093/nar/gkac1052}{Link}.
#'
#' @examples
#' \dontrun{
#'  SARSCoV2EmbeddingMatrix
#' }
"SARSCoV2EmbeddingMatrix"

#' Example Mapping for \emph{Severe Acute Respiratory Syndrome
#' Coronavirus 2 (SARS CoV-2)}
#'
#' A one-to-many mapping for \emph{SARS CoV-2} generated for illustrating how to
#' use the functions in this package.
#'
#' Mapping was generated by pairing proteins that have the same first 5
#' characters in their Uniprot Id. This mapping is strictly used for example
#' purposes.
#'
#' @format A data frame with 87 rows and 2 total columns. Columns Protein 1
#'     and Protein 2 are UniProt identifiers, and each row represents one
#'     pair of proteins with Uniprot Ids starting with the same 5 characters.
#'
#'
#' @examples
#' \dontrun{
#'  SARSCoV2Mapping
#' }
"SARSCoV2Mapping"

# [END]
