#' Raw Per-Protein Embeddings For Escherichia coli
#'
#' Raw per-protein embeddings generated by UniProt using the ProtT5 model
#' (prottrans_t5_xl_u50) for the reference proteome of Escherichia coli. Protein
#' sequences longer than 12k residues were excluded due to computational
#' limitations.
#'
#' @source UniProt.
#'
#' @format A data frame with 4,401 rows and 1024 total columns. In the index are
#'     UniProt identifiers for each protein, and columns are labelled V1, V2,
#'     ..., V1024 to represent each dimension of the embedding.
#' @examples
#' \dontrun{
#'  eColiEmbeddingMatrix
#' }
#' @source https://www.uniprot.org/help/downloads#embeddings
"eColiEmbeddingMatrix"

#' Paralog Mapping for Escherichia coli
#'
#' A one-to-many paralog mapping for Escherichia coli generated from the
#' Orthologous Matrix (OMA) Browser using the Genome Pair View tool with both
#' species as `Escherichia coli str. K-12 substr. MG1655 (ECOLI)`. Identifiers
#' were converted to UniProt using the Identifier Mapping file for download on
#' the OMA Browser. Processing was done to remove rows where either protein was
#' not present in the eColiEmbeddingMatrix to reduce the size of the file.
#'
#' @source Orthologous Matrix (OMA) Browser
#'
#' @format A data frame with 7,550 rows and 2 total columns. Columns Protein 1
#'     and Protein 2 are UniProt identifiers, and each row represents one
#'     paralogous pair.
#'
#' @examples
#' \dontrun{
#'  eColiParalogMapping
#' }
#' @source https://omabrowser.org/oma/genomePW/
"eColiParalogMapping"

# [END]
